File: src/test/java/org/radarbase/appserver/repository/UserRepositoryTest.java
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */

package org.radarbase.appserver.repository;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.radarbase.appserver.controller.FcmNotificationControllerTest.USER_ID;
import static org.radarbase.appserver.controller.RadarUserControllerTest.FCM_TOKEN_1;

import java.time.Instant;
import javax.persistence.PersistenceException;
import org.hibernate.exception.ConstraintViolationException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.runner.RunWith;
import org.radarbase.appserver.entity.Project;
import org.radarbase.appserver.entity.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
import org.springframework.test.context.junit4.SpringRunner;

@RunWith(SpringRunner.class)
@DataJpaTest
@EnableJpaAuditing
class UserRepositoryTest {

  @Autowired private transient TestEntityManager entityManager;

  @Autowired private transient UserRepository userRepository;

  private transient Project project;
  private transient Long projectId;
  private transient User user;
  private transient Long userId;

  @BeforeEach
  void setUp() {

    this.project = new Project().setProjectId("test-project");
    this.projectId = entityManager.persistAndGetId(project, Long.class);

    this.user =
        new User()
            .setFcmToken(FCM_TOKEN_1)
            .setEnrolmentDate(Instant.now())
            .setProject(project)
            .setTimezone(0d)
            .setLanguage("en")
            .setSubjectId(USER_ID);
    this.userId = entityManager.persistAndGetId(this.user, Long.class);
    entityManager.flush();
  }

  @Test
  public void whenInsertWithTransientProject_thenThrowException() {
    User user1 =
        new User()
            .setFcmToken(FCM_TOKEN_1)
            .setEnrolmentDate(Instant.now())
            .setProject(new Project())
            .setTimezone(0d)
            .setLanguage("en")
            .setSubjectId(USER_ID);

    IllegalStateException ex =
        assertThrows(
            IllegalStateException.class,
            () -> {
              entityManager.persist(user1);
              entityManager.flush();
            });

    assertTrue(ex.getMessage().contains("Not-null property references a transient value"));
  }

  @Test
  public void whenFindUserBySubjectId_thenReturnUser() {

    assertEquals(
        userRepository.findBySubjectId(USER_ID).get(), entityManager.find(User.class, this.userId));
  }

  @Test
  public void whenFindByProjectId_thenReturnUsers() {
    assertEquals(
        userRepository.findByProjectId(this.projectId).get(0),
        entityManager.find(User.class, this.userId));
  }

  @Test
  public void whenFindBySubjectIdAndProjectId_thenReturnUser() {
    assertEquals(
        userRepository.findBySubjectIdAndProjectId(USER_ID, this.projectId).get(),
        entityManager.find(User.class, this.userId));
  }

  @Test
  public void whenFindByFcmToken_thenReturnUser() {
    assertEquals(
        userRepository.findByFcmToken(FCM_TOKEN_1).get(), entityManager.find(User.class, this.userId));
  }

  @Test
  public void whenInsertWithExistingFcmToken_thenThrowException() {
    User user1 =
        new User()
            .setFcmToken(FCM_TOKEN_1)
            .setEnrolmentDate(Instant.now())
            .setProject(this.project)
            .setTimezone(0d)
            .setLanguage("en")
            .setSubjectId(USER_ID + "-2");

    PersistenceException ex =
        assertThrows(
            PersistenceException.class,
            () -> {
              entityManager.persistAndGetId(user1, Long.class);
              entityManager.flush();
            });

    assertEquals(ConstraintViolationException.class, ex.getCause().getClass());
  }
}

File: src/test/java/org/radarbase/appserver/repository/UserRepositoryTest.kt
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */
package org.radarbase.appserver.repository

import jakarta.persistence.PersistenceException
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.radarbase.appserver.controller.FcmDataMessageControllerTest.Companion.USER_ID
import org.radarbase.appserver.controller.RadarUserControllerTest.Companion.FCM_TOKEN_1
import org.radarbase.appserver.controller.RadarUserControllerTest.Companion.TIMEZONE
import org.radarbase.appserver.entity.Project
import org.radarbase.appserver.entity.User
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager
import org.springframework.data.jpa.repository.config.EnableJpaAuditing
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.sql.SQLIntegrityConstraintViolationException
import java.time.Instant


@ExtendWith(SpringExtension::class)
@DataJpaTest
@EnableJpaAuditing
class UserRepositoryTest {

    @Autowired
    private lateinit var entityManager: TestEntityManager

    @Autowired
    private lateinit var userRepository: UserRepository

    private lateinit var project: Project
    private var projectId: Long? = null
    private var userId: Long? = null

    @BeforeEach
    fun setUp() {
        this.project = Project(null, "test-project")
        this.projectId = entityManager.persistAndGetId(project) as Long

        val user = User(
            null, USER_ID, null, FCM_TOKEN_1, project,
            Instant.now(), null, TIMEZONE, "en", null
        )

        this.userId = entityManager.persistAndGetId(user) as Long
        entityManager.flush()
    }

    @Test
    fun `when insert with transient project then throw exception`() {
        val user1 = User(
            null, USER_ID, null, FCM_TOKEN_1, Project(),
            Instant.now(), null, TIMEZONE, "en", null
        )

        val ex = assertThrows<IllegalStateException> {
            entityManager.persist(user1)
            entityManager.flush()
        }

        assertTrue(ex.message!!.contains("Not-null property references a transient value"))
    }

    @Test
    fun `when find user by subject id then return user`() {
        assertEquals(userRepository.findBySubjectId(USER_ID), entityManager.find(User::class.java, this.userId))
    }

    @Test
    fun `when find by project id then return users`() {
        assertEquals(userRepository.findByProjectId(this.projectId!!)[0], entityManager.find(User::class.java, this.userId))
    }

    @Test
    fun `when find by subject id and project id then return user`() {
        assertEquals(
            userRepository.findBySubjectIdAndProjectId(USER_ID, this.projectId!!),
            entityManager.find(User::class.java, this.userId)
        )
    }

    @Test
    fun `when find by FCM token then return user`() {
        assertEquals(userRepository.findByFcmToken(FCM_TOKEN_1), entityManager.find(User::class.java, this.userId))
    }

    @Test
    fun `when insert with existing FCM token then throw exception`() {
        val user1 = User(
            null, "$USER_ID-2", null, FCM_TOKEN_1, this.project,
            Instant.now(), null, TIMEZONE, "en", null
        )

        val ex = assertThrows<PersistenceException> {
            entityManager.persistAndGetId(user1)
            entityManager.flush()
        }

        assertEquals(SQLIntegrityConstraintViolationException::class.java, ex.cause!!::class.java)
    }
}

